lahja/asyncio/endpoint.py

```diff
@@ -532,37 +535,56 @@ def connect_to_endpoints_nowait(self, *endpoints: ConnectionConfig) -> None:
 
         conn = await Connection.connect_to(config.path)
         remote = RemoteEndpoint(config.name, conn, self._receiving_queue.put)
-        self._full_connections[config.name] = remote
 
         task = asyncio.ensure_future(self._handle_server(remote))
-        subscriptions_task = asyncio.ensure_future(
-            self.watch_outbound_subscriptions(remote)
-        )
-        subscriptions_task.add_done_callback(self._endpoint_tasks.remove)
-        self._endpoint_tasks.add(subscriptions_task)
 
         task.add_done_callback(self._endpoint_tasks.remove)
-        task.add_done_callback(lambda _: subscriptions_task.cancel())
 
         self._endpoint_tasks.add(task)
 
-        # don't return control until the caller can safely call broadcast()
+        # block until we've received the other endpoint's subscriptions to
+        # ensure that it is safe to broadcast
         await remote.wait_until_subscription_received()
 
+        async with self._remote_connections_changed:
+            self._full_connections[config.name] = remote
+            self._remote_connections_changed.notify_all()
+
     async def _handle_server(self, remote: RemoteEndpoint) -> None:
         try:
             async with run_remote_endpoint(remote):
                 await remote.wait_stopped()
         finally:
             if remote.name is not None:
-                self._full_connections.pop(remote.name)
+                async with self._remote_connections_changed:
+                    self._full_connections.pop(remote.name)
+                    self._remote_connections_changed.notify_all()
+
+    async def wait_until_remote_subscriptions_change(self) -> None:
+        # We construct a set of awaitables that includes each individual
+        # remote's `wait_until_subscription_received` and a single call to the
+        # local `wait_until_connections_change`.  Once any one of these
+        # returns, we return, signaling that something has changed about our
+        # remote subscriptions.
+
+        wait_remote_subscriptions_change = tuple(
+            remote.wait_until_subscription_received()
+            for remote in self._full_connections.values()
+        )
+        wait_connection_changed = self.wait_until_connections_change()
 
-    async def watch_outbound_subscriptions(self, outbound: RemoteEndpoint) -> None:
-        while outbound in self._full_connections.values():
-            await outbound.wait_until_subscription_received()
-            await self.broadcast(
-                RemoteSubscriptionChanged(), BroadcastConfig(internal=True)
-            )
+        _, pending = await asyncio.wait(
+            (wait_connection_changed,) + wait_remote_subscriptions_change,
+            return_when=asyncio.FIRST_COMPLETED,
+        )
+
+        # We cancel the other pending waits
+        for task in pending:
+            task.cancel()
```

Comment:
I can't believe it's so hard to do this in asyncio, it feels like this should be a common pattern but I remember ending up at the same place.

Reply:
If it's any solitude, `trio` doesn't have any equivalent to `asyncio.gather`.

You can see my ugly stab at implementing it here: https://github.com/ethereum/trinity/pull/615/files#diff-32f4c7ca496f37029fb4feaa33835db1R38